Outliner _ 아웃라이너 오브젝트 생성

sanghun219 opened the pull request:
#세부사항
1. Material의 수정 및 오브젝트 copy
  -  https://www.youtube.com/watch?v=BWoVkQjYncY 유튜브를 보고 material의 외곽선 부분만 색을 칠하는 방법을 참고
     기본적인 outline 생성 방식은 해당하는 actor를 복사하고 위의 material을 따로 적용시켜 원하는 색으로 나타내는것이다.
     원래는 mouse나 vr기기를 통한 포인터에 의해 outline이 켜지고 꺼져야하지만 외부 모듈이기에 1,2 버튼으로 편리하게
      결과를 확인하게 했음.
2. outline Actor 삭제
 -  outline이 존재한다는 조건은 해당 actor의 child중에 OutLiner라는 이름을 가지는 actor를 찾아 있다면 생성을 불가능하게
    없다면 생성이 가능하게 하였고 , 삭제는 actor를 찾아 없다면 삭제를 불가능하게 있다면 삭제를 가능하게 하였다.
    LastOutlineActor는 생성할 때마다 해당 actor로 변경해주고 삭제해 줄 때 nullptr 처리를 하게 하였다.
※3번 버튼은 LastOutlineActor의 Label을 띄워주는 기능으로 넣어둠.

bangbangdog219 committed:
Outliner _ 아웃라이너 오브젝트 생성

#세부사항
1. Material의 수정 및 오브젝트 copy
  -  https://www.youtube.com/watch?v=BWoVkQjYncY 유튜브를 보고 material의 외곽선 부분만 색을 칠하는 방법을 참고
     기본적인 outline 생성 방식은 해당하는 actor를 복사하고 위의 material을 따로 적용시켜 원하는 색으로 나타내는것이다.
     원래는 mouse나 vr기기를 통한 포인터에 의해 outline이 켜지고 꺼져야하지만 외부 모듈이기에 1,2 버튼으로 편리하게
      결과를 확인하게 했음.
2. outline Actor 삭제
 -  outline이 존재한다는 조건은 해당 actor의 child중에 OutLiner라는 이름을 가지는 actor를 찾아 있다면 생성을 불가능하게
    없다면 생성이 가능하게 하였고 , 삭제는 actor를 찾아 없다면 삭제를 불가능하게 있다면 삭제를 가능하게 하였다.
    LastOutlineActor는 생성할 때마다 해당 actor로 변경해주고 삭제해 줄 때 nullptr 처리를 하게 하였다.
※3번 버튼은 LastOutlineActor의 Label을 띄워주는 기능으로 넣어둠.

Minusi commented on UEPrototype/Source/UEPrototype/Private/Interaction/Outliner.cpp:
```diff
@@ -0,0 +1,83 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#include "Outliner.h"
+#include "../Public/Interaction/Outliner.h"
```
자동적으로 생성된 ../Public/..(이후 생략)은 다시 재작성해서 다음과 같이 작성하는 게 직관성에 좋을 것 같음 :

#include "Interaction/Outliner.h"

Minusi commented on UEPrototype/Source/UEPrototype/Public/Interaction/Outliner.h:
```diff
@@ -0,0 +1,39 @@
+	FORCEINLINE AActor* GetLastOutlinedActor()const { return LastOutlinedActor; }
+
+	UFUNCTION(BlueprintCallable, Category ="Outliner")
+	void DrawActorOutline(const AActor* actor);
```
매개변수 이름이 카멜 케이싱을 따르고 있으므로 지정된 코드 컨벤션 규칙을 위배하고 있음.
Actor로 고치거나 또는 적합한 이름의 매개변수로 첫 문자를 대문자화하여 수정해주셈.

Minusi commented on UEPrototype/Source/UEPrototype/Public/Interaction/Outliner.h:
```diff
@@ -0,0 +1,39 @@
+
+	const FString OutlinerLabelName = "OutLiner";
+
+};
```
클래스부터 시작해서 멤버 변수 및 함수에 대해서 주석을 꼭 달아주길 바람. 전부 어떤 기능을 하는 건지에 대해서 상세한 주석 작성해주셈.

Minusi commented on UEPrototype/Config/DefaultEngine.ini:
```diff
@@ -111,7 +111,7 @@ DefaultGraphicsPerformance=Scalable
 
 [/Script/EngineSettings.GameMapsSettings]
-EditorStartupMap=/Game/Map/Map_Prototype.Map_Prototype
+EditorStartupMap=/Game/Map/Test_Sanghun.Test_Sanghun
```
프로젝트 설정에서 Map_Prototype으로 다시 변경하는 것을 권장함. 개인 작업이 마스터의 시작 맵에 영향을 주지 않도록 하는 걸로.

Minusi commented on UEPrototype/Config/DefaultEngine.ini:
```diff
@@ -120,7 +120,7 @@ ThreePlayerSplitscreenLayout=FavorTop
 bOffsetPlayerGamepadIds=False
 GameInstanceClass=/Script/Engine.GameInstance
-GameDefaultMap=/Game/Map/Map_Prototype.Map_Prototype
+GameDefaultMap=/Game/Map/Test_Sanghun.Test_Sanghun
```
프로젝트 설정에서 Map_Prototype으로 다시 변경하는 것을 권장함. 개인 작업이 마스터의 시작 맵에 영향을 주지 않도록 하는 걸로.

Minusi reviewed (changes requested):
5가지에 대한 코드 리뷰 피드백 제출함. 읽고 반영해주고 곧바로 이 모듈에서 추가 작업을 진행할 것 같으니, 수정만 하고 바로 pull request하지 마셈.

bangbangdog219 committed:
VPAmbientSound _ Interaction module 생성 및 VPAmbientSound 생성, Outliner의 위치를 Interaction Module로 옮김

#세부사항
1. Outliner를 담고있던 Interface 모듈을 삭제하고 ActorInfo 모듈 생성 및 Outliner를 이 폴더로 옮김
2. ActorInfo 모듈에 VPAmbientSound 클래스 생성
  - 사용자에게 제공되어야하는 프로퍼티를 구조체로 생성후 외부의 설정값을 받아 AudioComponent 값을 변경시킴
  - 너무 디테일한 프로퍼티에 대해서는 AmbientSound를 상속하면 자동으로 default값으로 설정을 해준다.
  - 조정되어야할 default값(기본적으로 sound 볼륨이 0으로되어있던점, 사운드 적용 거리가 너무 큰 점)을 CreateSound 함수 에서 혹은 생성자에서 처리를 해 준다.
  - 기본적으로 제공되어야 하는 최소한의 기능 (play,pause,stop,adjustVolume)을 제공함, 추후 필요하다 싶으면 함수를 추가하여 사용
 - 사운드는 각 액터당 하나 씩 생성할 수 있도록 설정함.
 - 5번이 사운드 생성 , 8번이 pause , 9번이 stop으로 설정 되어있음.

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+#include <UObjectGlobals.h>
+#include <Materials/MaterialInstanceDynamic.h>
+
+UOutliner::UOutliner()
```
사용되지 않는 생성자 삭제할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+		return;
+	}
+
+	//Actor�� ����ϴ� ���. ����̱� ������ spawn����� �ʿ��� parameter�� ��� �ش� Actor�� ���� �͵�.
```
주석 꺠짐 : ㅁㅁ문에라고 뜨는 것 수정할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+	AActor* SpawnedOutlineActor = GetWorld()->SpawnActor<AActor>(Actor->GetClass(), Actor->GetActorLocation(), Actor->GetActorRotation());
+
+	// �ƿ����� ������ ������ �ڽ���� �־��ش�(����� Ʈ������ ���� �̵��� ���ؼ� ���� ��󰡾��ϴϱ�)
+	SpawnedOutlineActor->AttachToActor(const_cast<AActor*>(Actor), FAttachmentTransformRules::KeepWorldTransform);
```
const_cast를 제거하고 매개변수의 const 키워드를 제거하거나 다른 방법을 모색할 것. 잘못된 설계

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+	//�׸��� �ߺ� �˻�, LableName� �˻��� �ش� �ڽ� Actor�� LabelName�� �ִٸ� ����Ѵ�.
+	TArray<AActor*> ChildOutliner;
+	Actor->GetAttachedActors(ChildOutliner);
+	for (int i = 0; i < ChildOutliner.Num(); ++i)
```
반복자를 통한 순회로 재구성하고 반복자에 상수성을 부과할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+
+	//Material� ����ϴ� ���. �ش� ������ ��� Component��� ��ȸ�Ͽ� StaticMeshComponent�� �ִٸ� �� Component�� material�
+	//�̸� ������� material�� ��ü��.
+	for (int i = 0; i < SM_Comp.Num(); i++)
```
반복자를 통한 순회로 재구성하고 반복자에 상수성을 부과할 것

Minusi replied:
반복자를 통한 순회로 재구성하고 반복자에 상수성을 부과할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/Outliner.cpp:
```diff
@@ -0,0 +1,99 @@
+	if (ChildOutliner.Num() == 0) return;
+
+	//�ڽ� actor��� ��ȸ�Ͽ� �ش� LableName�� �ִٸ� ����� ������ش�.
+	for (int i = 0; i < ChildOutliner.Num(); ++i)
```
반복자를 통한 순회로 재구성하고 반복자에 상수성을 부과할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+		= FVector(400.f, 0.0f, 0.0f);
+}
+
+void AVPAmbientSound::CreateSound(const AActor* Actor, FSoundProperty SoundProperty, FAttenuationProperty AttenuationProperty)
```
Create라는 이름보다는 Initialize 또는 그에 준하는 문맥에 맞는 네이밍으로 대체할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+		//��ȭ ��뿡 ���� bool��(�ڷᰡ ���)
+		GetAudioComponent()->AttenuationOverrides.bSpatialize = true;
+
+		//2���� ���� �ݰ�, 3���� ����� ���� 0��� ���
```
주석 깨짐 현상 수정할 것 : 'ㅁㅁ는 0으로 고정'

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+
+	{
+		//��ȭ�� �� ���ΰ�(�ϴ� ����̱��ѵ� ����ڿ��� �Ѱ����� �ȳѰ����� ���غ����)
+		GetAudioComponent()->bAllowSpatialization = true;
```
반복되는 Get..Component()를 호출할 때에는 이것을 한 번 호출하여 포인터에 담아두고 캐시된 포인터를 이용하여 접근할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+	this->SetActorLocation(Actor->GetActorLocation());
+
+	//�ش� Actor�� �ٿ��� Actor�� ���̰� ������� Ʈ������� �����.
+	this->AttachToActor(const_cast<AActor*>(Actor), FAttachmentTransformRules::KeepWorldTransform);
```
UOutliner에서 설명했듯이 const_cast의 사용은 일반적으로 설계가 잘못되었음을 드러내는 상황임. 매개변수의 const성을 제거하거나 Actor의 상수성을 보존할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+		Play();
+}
+//���� Actor ���
+void AVPAmbientSound::DestroySound()
```
Destory 래퍼 함수의 큰 의미가 없어보이므로 Actor의 Destroy 일관성을 위해 삭제할 것

Minusi replied:
블루프린트 노출용 래퍼함수라고 들었는데, 확인하고 기존 함수가 BP성을 제공한다면 삭제할 것.

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+	Destroy();
+}
+//���̵� ���� ��� ���ص� �ɵ�
+void AVPAmbientSound::FadeInVP(float FadeInDuration, float FadeVolumeLevel)
```
VP라는 네이밍은 일반적으로 프로젝트의 축약어로 쓰이는 용도이므로, 앰비언트 사운드 관련 클래스에서 이 네이밍을 사용하는 것은 부적절해 보임. 좀 더 이해할 수 있는 이름으로 수정할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+	FadeIn(FadeInDuration, FadeVolumeLevel);
+}
+
+void AVPAmbientSound::FadeOutVP(float FadeOutDuration, float FadeVolumeLevel)
```
VP라는 네이밍은 일반적으로 프로젝트의 축약어로 쓰이는 용도이므로, 앰비언트 사운드 관련 클래스에서 이 네이밍을 사용하는 것은 부적절해 보임. 좀 더 이해할 수 있는 이름으로 수정할 것

Minusi replied:
하위의 모든 VP 네이밍 함수들에 이를 적용할 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+
+
+
+void AVPAmbientSound::SetupGizmo(FTransform Transform)
```
이거 설계와 관련 없는 부분이므로 지울 것

Minusi commented on UEPrototype/Source/UEPrototype/Private/ActorInfo/VPAmbientSound.cpp:
```diff
@@ -0,0 +1,201 @@
+{
+	TArray<AActor*> ChildSound;
+	Actor->GetAttachedActors(ChildSound);
+	for (int i = 0; i < ChildSound.Num(); ++i)
```
마찬가지로 반복자 순회를 사용할 것

Minusi commented on UEPrototype/Source/UEPrototype/Public/ActorInfo/VPAmbientSound.h:
```diff
@@ -0,0 +1,169 @@
+#include <Components/AudioComponent.h>
+#include <Sound/SoundAttenuation.h>
+#include "VPAmbientSound.generated.h"
+/**
```
필요없는 주석 블록 삭제할 것

Minusi reviewed (changes requested):
수정하고 수정되었음을 표시할 것

bangbangdog219 committed:
SoundEdit _ AVPAmbientSound 수정

#세부사항
1. VP함수들 제거
  - AmbientSound에 존재하는 함수들은 다시 만들지않고 기존의 것들을 사용
  -불필요한 VP 네이밍 제거
2. for문들 iterator로 수정
3.Pause() 제거
 - AudioComponent의 SetPause() 사용, 나중에 설정 값을 대비해서 bPause를 남겨둠

bangbangdog219 committed:
SoundEdit _ AVPAmbientSound 수정

# 세부사항
 1. VP관련 함수 제거
   - AmbientSound에 정의되어 있는 함수들은 그대로 사용하기로 함.
   - VP 네이밍 수정

bangbangdog219 committed:
merge - 충돌 처리 테스트

bangbangdog219 committed:
CommitError - 맵 프로토타입 기존의 것으로 대체함.

Minusi committed:
Merge branch 'master' into Interaction
